src/main/java/org/jenkinsci/plugins/workflow/cps/nodes/StepDescriptorCache.java

```diff
@@ -55,26 +56,30 @@ public static void invalidateGlobalCache() {
     }
 
     public void invalidateAll() {
-        descriptorCache.clear();
+        store.clear();
     }
 
-    private final Memoizer<String, StepDescriptor> descriptorCache = new Memoizer<String, StepDescriptor>() {
-
-        public StepDescriptor compute(String descriptorId) {
-            Jenkins j = Jenkins.getInstance();
-            if (descriptorId != null && j != null) {
-                return (StepDescriptor) j.getDescriptor(descriptorId);
-            }
-            return null;
-        }
-    };
+    private final ConcurrentHashMap<String, StepDescriptor> store = new ConcurrentHashMap<String, StepDescriptor>();
 
     @CheckForNull
     public StepDescriptor getDescriptor(String descriptorId) {
-        if (descriptorId != null) {
-            return descriptorCache.get(descriptorId);
+        if (descriptorId == null) {
+            return null;
         }
-        return null;
 
+        StepDescriptor v = store.get(descriptorId);
+        if(v!=null) return v;
+
+        synchronized (this) {
```

Comment:
The code in `Memoizer` is not good. It is _one_ (bad) workaround for the case where there is a problem if two threads compute the same value, because the cache function is not idempotent (or is so expensive that we would rather block another thread than do it twice). That is not the case here: the lookup is quick and harmless to do in two different threads in the unlikely event they overlap. So you should remove the redundant check and the synchronization.